src/Parser/TypeParser.php

```diff
@@ -468,53 +470,76 @@ private function tryParseArrayOrOffsetAccess(TokenIterator $tokens, Ast\Type\Typ
 		return $type;
 	}
 
-
 	/** @phpstan-impure */
-	private function parseArrayShape(TokenIterator $tokens, Ast\Type\TypeNode $type): Ast\Type\ArrayShapeNode
+	private function parseShape(TokenIterator $tokens, Ast\Type\IdentifierTypeNode $type): array
 	{
 		$tokens->consumeTokenType(Lexer::TOKEN_OPEN_CURLY_BRACKET);
 		if ($tokens->tryConsumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET)) {
-			return new Ast\Type\ArrayShapeNode([]);
+			return [];
 		}
 
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
-		$items = [$this->parseArrayShapeItem($tokens)];
+		$items = [$this->parseShapeItem($tokens, $type)];
 
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 		while ($tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA)) {
 			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 			if ($tokens->tryConsumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET)) {
 				// trailing comma case
-				return new Ast\Type\ArrayShapeNode($items);
+				return $items;
 			}
 
-			$items[] = $this->parseArrayShapeItem($tokens);
+			$items[] = $this->parseShapeItem($tokens, $type);
 			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 		}
 
 		$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
 		$tokens->consumeTokenType(Lexer::TOKEN_CLOSE_CURLY_BRACKET);
 
+		return $items;
+	}
+
+	/** @phpstan-impure */
+	private function parseArrayShape(TokenIterator $tokens, Ast\Type\IdentifierTypeNode $type): Ast\Type\ArrayShapeNode
+	{
+		$items = $this->parseShape($tokens, $type);
+
 		return new Ast\Type\ArrayShapeNode($items);
 	}
 
+	/** @phpstan-impure */
+	private function parseObjectShape(TokenIterator $tokens, Ast\Type\IdentifierTypeNode $type): Ast\Type\ObjectShapeNode
+	{
+		$items = $this->parseShape($tokens, $type);
+
+		return new Ast\Type\ObjectShapeNode($type, $items);
+	}
+
 
 	/** @phpstan-impure */
-	private function parseArrayShapeItem(TokenIterator $tokens): Ast\Type\ArrayShapeItemNode
+	private function parseShapeItem(TokenIterator $tokens, Ast\Type\IdentifierTypeNode $type): Ast\Type\ShapeItemNode
 	{
 		try {
 			$tokens->pushSavePoint();
-			$key = $this->parseArrayShapeKey($tokens);
+			$key = $this->parseShapeKey($tokens);
 			$optional = $tokens->tryConsumeTokenType(Lexer::TOKEN_NULLABLE);
 			$tokens->consumeTokenType(Lexer::TOKEN_COLON);
 			$value = $this->parse($tokens);
 			$tokens->dropSavePoint();
 
+			if ($type->name === 'object') {
```

Comment:
not quite sure on this approach to re-using the `parseArrayShapeItem` code, it seems fairly brittle but the other option would be to `return [$key, $optionnal, $value];` and then construct the correct value in either `parseArrayShape` or `parseObjectShape`, but that felt pretty messy too.